src/tasks/dump.ts

```diff
@@ -0,0 +1,567 @@
+
+import { estimateTradeAmount, getFee, placeOrder } from "../services/api";
+import {
+  BUY_ETH_ADDRESS,
+  domain,
+  Order,
+  OrderKind,
+  SigningScheme,
+  signOrder,
+  TypedDataDomain,
+} from "../ts";
+
+import {
+  getDeployedContract,
+  isSupportedNetwork,
+  SupportedNetwork,
+} from "./ts/deployment";
+import { promiseAllWithRateLimit } from "./ts/rate_limits";
+import { Align, displayTable } from "./ts/table";
+import {
+  isNativeToken,
+  nativeToken,
+  NativeToken,
+  tokenDetails,
+  TokenDetails,
+  WRAPPED_NATIVE_TOKEN_ADDRESS,
+  balanceOf,
+  transfer,
+  displayName,
+} from "./ts/tokens";
+import { formatTokenValue } from "./ts/value";
+
+const MAX_LATEST_BLOCK_DELAY_SECONDS = 2 * 60;
+const MAX_ORDER_VALIDITY_SECONDS = 24 * 3600;
+
+interface DumpInstruction {
+  token: TokenDetails;
+  amount: BigNumber;
+  approvedAmount: BigNumber;
+  receivedAmount: BigNumber;
+  balance: BigNumber;
+  fee: BigNumber;
+}
+
+interface DisplayDumpInstruction {
+  fromSymbol: string;
+  fromAddress: string;
+  balance: string;
+  needsAllowance: "yes" | "";
+  receivedAmount: string;
+  feePercent: string;
+}
+
+interface TransferToReceiver {
+  token: TokenDetails | NativeToken;
+  amount: BigNumber;
+}
+
+interface DumpInstructions {
+  instructions: DumpInstruction[];
+  toToken: TokenDetails | NativeToken;
+  // Amount of toToken that will be transfered to the receiver address with a
+  // standard erc20 transfer. This is only set if there is a receiver and the
+  // specified toToken is also a token to be dumped.
+  transferToReceiver?: TransferToReceiver;
+}
+
+interface GetDumpInstructionInput {
+  dumpedTokens: string[];
+  toTokenAddress: string | undefined; // undefined defaults to native token (e.g., ETH)
+  user: string;
+  allowanceManager: string;
+  maxFeePercent: number;
+  hasCustomReceiver: boolean;
+  hre: HardhatRuntimeEnvironment;
+  network: SupportedNetwork;
+}
+async function getDumpInstructions({
+  dumpedTokens: inputDumpedTokens,
+  toTokenAddress,
+  user,
+  allowanceManager,
+  maxFeePercent,
+  hasCustomReceiver,
+  hre,
+  network,
+}: GetDumpInstructionInput): Promise<DumpInstructions> {
+  // todo: support dumping ETH by wrapping them
+  if (inputDumpedTokens.includes(BUY_ETH_ADDRESS)) {
+    throw new Error(
+      `Dumping the native token is not supported. Remove the ETH flag address ${BUY_ETH_ADDRESS} from the list of tokens to dump.`,
+    );
+  }
+
+  const dumpedTokens = Array.from(new Set(inputDumpedTokens));
+  const toToken =
+    toTokenAddress === undefined || toTokenAddress === BUY_ETH_ADDRESS
+      ? nativeToken(hre)
+      : await tokenDetails(toTokenAddress, hre);
+
+  let transferToReceiver: TransferToReceiver | undefined = undefined;
+  const sameTokenIndex = dumpedTokens.findIndex(
+    (token) => token === (toTokenAddress ?? BUY_ETH_ADDRESS),
+  );
+  if (sameTokenIndex !== -1) {
+    dumpedTokens.splice(sameTokenIndex, 1);
+    if (hasCustomReceiver) {
```

Comment:
Is this if statement supposed to be nested? I'd think that we should also support custom receivers if `sameTokenIndex === -1`?

Edit: I think I see it now, this is only if people put the target token into the sell list. Is this really a feature we need? It feels like it adds a good amount of logic and is not really that useful (it requires an extra transfer that could also be done manually). We could in this case just fail and ask people to provide a sell list without the target token.

Reply:
Do you think that the added unit tests are enough to offset the extra logic for treating this extra case? I believe this feature will be helpful when automating the withdraw process, otherwise this case should be handled somewhere else.